`__setitem__` should accept a dictionary

**Describe the bug**
`__setitem__` does not accept a dictionary

**To Reproduce**
```python
import pandas as pd

df = pd.DataFrame(columns=["X"])
df.loc[0] = {"X": 0}
print(df)
```

```
>mypy --version
mypy 1.15.0 (compiled: yes)

>mypy bug.py
bug.py:4: error: No overload variant of "__setitem__" of "_LocIndexerFrame" matches argument types "int", "dict[str, int]"  [call-overload]
bug.py:4: note: Possible overload variants:
bug.py:4: note:     def [ScalarT: str | bytes | date | datetime | timedelta | <7 more items> | complex] __setitem__(self, Series[builtins.bool] | ndarray[tuple[int, ...], dtype[numpy.bool[builtins.bool]]] | list[builtins.bool] | str | str_ | <9 more items>, str | bytes | date | datetime | timedelta | <15 more items> | None, /) -> None
bug.py:4: note:     def __setitem__(self, tuple[tuple[Index[Any] | Series[builtins.bool] | ndarray[tuple[int, ...], dtype[numpy.bool[builtins.bool]]] | list[builtins.bool] | str | bytes | date | datetime | timedelta | <7 more items> | complex | list[Any] | slice[Any, Any, Any] | tuple[str | bytes | date | datetime | timedelta | <7 more items> | complex, ...], ...], Hashable], str | bytes | date | datetime | timedelta | <14 more items> | None, /) -> None
Found 1 error in 1 file (checked 1 source file)
```

```
>pyright --version
pyright 1.1.400

>pyright bug.py
bug.py
  bug.py:4:1 - error: No overloads for "__setitem__" match the provided arguments (reportCallIssue)
  bug.py:4:1 - error: Argument of type "dict[str, int]" cannot be assigned to parameter "value" of type "Scalar | ArrayLike | NAType | NaTType | Series[Unknown] | DataFrame | list[Unknown] | None" in function "__setitem__"
    Type "dict[str, int]" is not assignable to type "Scalar | ArrayLike | NAType | NaTType | Series[Unknown] | DataFrame | list[Unknown] | None"
      "dict[str, int]" is not assignable to "str"
      "dict[str, int]" is not assignable to "bytes"
      "dict[str, int]" is not assignable to "date"
      "dict[str, int]" is not assignable to "datetime"
      "dict[str, int]" is not assignable to "timedelta"
      "dict[str, int]" is not assignable to "datetime64[date | int | None]"
      "dict[str, int]" is not assignable to "timedelta64[timedelta | int | None]" (reportArgumentType)
2 errors, 0 warnings, 0 informations
```

```
>python --version
Python 3.13.2
```

```
>uv pip freeze | findstr "pandas"
pandas==2.2.3
pandas-stubs==2.2.3.250308
```

--- a/pandas-stubs/core/frame.pyi
+++ b/pandas-stubs/core/frame.pyi
@@ -177,7 +177,14 @@ class _iLocIndexerFrame(_iLocIndexer, Generic[_T]):
             | tuple[int, IndexType]
         ),
         value: (
-            Scalar | Series | DataFrame | np.ndarray | NAType | NaTType | dict | None
+            Scalar
+            | Series
+            | DataFrame
+            | np.ndarray
+            | NAType
+            | NaTType
+            | Mapping[Hashable, Scalar | NAType | NaTType]
+            | None
         ),
     ) -> None: ...
 
@@ -249,7 +256,7 @@ class _LocIndexerFrame(_LocIndexer, Generic[_T]):
             | Series
             | DataFrame
             | list
-            | dict
+            | Mapping[Hashable, Scalar | NAType | NaTType]
             | None
         ),
     ) -> None: ...

--- a/pandas-stubs/core/frame.pyi
+++ b/pandas-stubs/core/frame.pyi
@@ -176,7 +176,9 @@ class _iLocIndexerFrame(_iLocIndexer, Generic[_T]):
             | tuple[IndexType, IndexType]
             | tuple[int, IndexType]
         ),
-        value: Scalar | Series | DataFrame | np.ndarray | NAType | NaTType | None,
+        value: (
+            Scalar | Series | DataFrame | np.ndarray | NAType | NaTType | dict | None
+        ),
     ) -> None: ...
 
 class _LocIndexerFrame(_LocIndexer, Generic[_T]):
@@ -239,13 +241,23 @@ class _LocIndexerFrame(_LocIndexer, Generic[_T]):
         idx: (
             MaskType | StrLike | _IndexSliceTuple | list[ScalarT] | IndexingInt | slice
         ),
-        value: Scalar | NAType | NaTType | ArrayLike | Series | DataFrame | list | None,
+        value: (
+            Scalar
+            | NAType
+            | NaTType
+            | ArrayLike
+            | Series
+            | DataFrame
+            | list
+            | dict
+            | None
+        ),
     ) -> None: ...
     @overload
     def __setitem__(
         self,
         idx: tuple[_IndexSliceTuple, Hashable],
-        value: Scalar | NAType | NaTType | ArrayLike | Series | list | None,
+        value: Scalar | NAType | NaTType | ArrayLike | Series | list | dict | None,
     ) -> None: ...
 
 # With mypy 1.14.1 and python 3.12, the second overload needs a type-ignore statement

--- a/tests/test_frame.py
+++ b/tests/test_frame.py
@@ -4134,3 +4134,13 @@ def test_combine() -> None:
         ),
         pd.DataFrame,
     )
+
+
+def test_df_loc_dict() -> None:
+    """Test that we can set a dict to a df.loc result GH1203."""
+    df = pd.DataFrame(columns=["X"])
+    df.loc[0] = {"X": 0}
+    check(assert_type(df, pd.DataFrame), pd.DataFrame)
+
+    df.iloc[0] = {"X": 0}
+    check(assert_type(df, pd.DataFrame), pd.DataFrame)

